[MRG] FEA Add support for Individual Conditional Expectation (ICE) plots (#16619)
scikit-learn/scikit-learn: doc/modules/partial_dependence.rst

feature, an ICE plot visualizes the dependence of the prediction on a
feature for each sample separately with one line per sample.
Due to the limits of human perception, only one input feature of interest is
supported for ICE plots.

The figures below show four ICE plots for the California housing dataset,
with a :class:`HistGradientBoostingRegressor
<sklearn.ensemble.HistGradientBoostingRegressor>`. The second figure plots
the corresponding PD line overlaid on ICE lines.

.. figure:: ../auto_examples/inspection/images/sphx_glr_plot_partial_dependence_002.png
   :target: ../auto_examples/inspection/plot_partial_dependence.html
   :align: center
   :scale: 70

While the PDPs are good at showing the average effect of the target features,
they can obscure a heterogeneous relationship created by interactions.
When interactions are present the ICE plot will provide many more insights.
For example, we could observe a linear relationship between the median income
and the house price in the PD line. However, the ICE lines show that there
are some exceptions, where the house price remains constant in some ranges of

> the median income.

Reviewer: There are no PDP plots with MedInc anymore

The :mod:`sklearn.inspection` module's :func:`plot_partial_dependence`
convenience function can be used to create ICE plots by setting
``kind='individual'``. In the example below, we show how to create a grid of
ICE plots:

    >>> from sklearn.datasets import make_hastie_10_2
    >>> from sklearn.ensemble import GradientBoostingClassifier
    >>> from sklearn.inspection import plot_partial_dependence

    >>> X, y = make_hastie_10_2(random_state=0)
    >>> clf = GradientBoostingClassifier(n_estimators=100, learning_rate=1.0,
    ...     max_depth=1, random_state=0).fit(X, y)
    >>> features = [0, 1]
    >>> plot_partial_dependence(clf, X, features,
    ...     kind='individual')  # doctest: +SKIP

In ICE plots it might not be easy to see the average effect of the input
feature of interest. Hence, it is recommended to use ICE plots alongside

> partial dependency plots. They can be plotted together with

Reviewer: ```suggestion
partial dependence plots. They can be plotted together with
```

with a :class:`GradientBoostingRegressor
<sklearn.ensemble.GradientBoostingRegressor>`. The second figure plots
the corresponding PD line overlaid on ICE lines.

.. figure:: ../auto_examples/inspection/images/sphx_glr_plot_partial_dependence_004.png
   :target: ../auto_examples/inspection/plot_partial_dependence.html
   :align: center
   :scale: 70

.. figure:: ../auto_examples/inspection/images/sphx_glr_plot_partial_dependence_005.png
   :target: ../auto_examples/inspection/plot_partial_dependence.html
   :align: center
   :scale: 70

While the PDPs are good at showing the average effect of the target features,
they can obscure a heterogeneous relationship created by interactions.
When interactions are present the ICE plot will provide many more insights.
For example, we could observe a linear relationship between the median income
and the house price in the PD line. However, the ICE lines show that there
are some exceptions, where the house price remains constant in some ranges of
the median income.
 
The :mod:`sklearn.inspection` module's :func:`plot_partial_dependence`
convenience function can be used to create ICE plots by setting

> ``individual=True``. In the example below, we show how to create a grid of

Reviewer 2: We could add that we can set `individual="both"` to both ICE and PD. In addition, we can control the number of ICE using the `subsample` parameter.

Reviewer 2: Oh I see now that individual="both" is discussed below.

ICE plots:

>     >>> from sklearn.datasets import make_hastie_10_2

Reviewer 2: I would remove the example and rather link the real example

Author: @glemaitre Do you mean linking to https://97096-843222-gh.circle-artifacts.com/0/doc/auto_examples/inspection/plot_partial_dependence.html? It is anyway linked at the end of the page.

Reviewer 2: Oh sorry I thought I removed this comment. Keep the example since we have one in the above section in a similar manner. At first, I thought we did not have one above.

Author: No problem. I'll mark this as resolved.

    samples differently. Remember, however, that the primary assumption for
    interpreting PDPs is that the features should be independent.

Individual conditional expectation (ICE) plot
=============================================

Similar to a PDP, an individual conditional expectation (ICE) plot

> shows the dependence between the target function and a target feature.

Author: Done in https://github.com/scikit-learn/scikit-learn/pull/16619/commits/72c474d950e146780962e594cd387df7c46b852d

with a :class:`GradientBoostingRegressor
<sklearn.ensemble.GradientBoostingRegressor>`. The second figure plots
the corresponding PD line overlaid on ICE lines.

.. figure:: ../auto_examples/inspection/images/sphx_glr_plot_partial_dependence_004.png
   :target: ../auto_examples/inspection/plot_partial_dependence.html
   :align: center
   :scale: 70

.. figure:: ../auto_examples/inspection/images/sphx_glr_plot_partial_dependence_005.png
   :target: ../auto_examples/inspection/plot_partial_dependence.html
   :align: center
   :scale: 70

While the PDPs are good at showing the average effect of the target features,
they can obscure a heterogeneous relationship created by interactions.
When interactions are present the ICE plot will provide many more insights.
For example, we could observe a linear relationship between the median income
and the house price in the PD line. However, the ICE lines show that there
are some exceptions, where the house price remains constant in some ranges of
the median income.

Similar to PDPs, the target features are assumed to be independent from the
complement features. Thus, in the case of correlated features, we will create absurd

> data points to compute the ICE.

Author: Done in https://github.com/scikit-learn/scikit-learn/pull/16619/commits/a36fd86d3b74b1f422ad4a1088fbec16a0a0eb0f

    samples differently. Remember, however, that the primary assumption for
    interpreting PDPs is that the features should be independent.

Individual conditional expectation (ICE) plot
=============================================

Similar to a PDP, an individual conditional expectation (ICE) plot
shows the dependence between the target function and a target feature.
However, unlike a PDP, which shows the average effect of the target
feature, an ICE plot visualizes the dependence of the prediction on a

> feature for each instance separately with one line per instance.

Reviewer 2: ```suggestion
feature for each sample separately with one line per sample.
```

Due to the limits of human perception, only one target feature is
supported for ICE plots.

The figures below show four ICE plots for the California housing dataset,
with a :class:`GradientBoostingRegressor
<sklearn.ensemble.GradientBoostingRegressor>`. The second figure plots
the corresponding PD line overlaid on ICE lines.

.. figure:: ../auto_examples/inspection/images/sphx_glr_plot_partial_dependence_004.png
   :target: ../auto_examples/inspection/plot_partial_dependence.html
   :align: center
   :scale: 70

.. figure:: ../auto_examples/inspection/images/sphx_glr_plot_partial_dependence_005.png
   :target: ../auto_examples/inspection/plot_partial_dependence.html
   :align: center
   :scale: 70

While the PDPs are good at showing the average effect of the target features,
they can obscure a heterogeneous relationship created by interactions.

> When interactions are present the ICE plot will provide much more insight.

Reviewer 2: ```suggestion
When interactions are present the ICE plot will provide many more insights.
```

For example, we could observe a linear relationship between the median income
and the house price in the PD line. However, the ICE lines show that there
are some exceptions, where the house price remains constant in some ranges of
the median income.

Similar to PDPs, the target features are assumed to be independent from the

> complement features.

Reviewer 2: I would add what is the inconvenience of this assumption:

```suggestion
complement features. Thus, in the case of correlated features, we will create absurd
data points to compute the ICE.
 
```

The :mod:`sklearn.inspection` module's :func:`plot_partial_dependence`
convenience function can be used to create ICE plots by setting

> ``individual=True``. In the below example we show how to create a grid of

Reviewer 2: ```suggestion
``individual=True``. In the example below, we show how to create a grid of
```

ICE plots:

    >>> from sklearn.datasets import make_hastie_10_2
    >>> from sklearn.ensemble import GradientBoostingClassifier
    >>> from sklearn.inspection import plot_partial_dependence

    >>> X, y = make_hastie_10_2(random_state=0)
    >>> clf = GradientBoostingClassifier(n_estimators=100, learning_rate=1.0,
    ...     max_depth=1, random_state=0).fit(X, y)
    >>> features = [0, 1]
    >>> plot_partial_dependence(clf, X, features,

>     ...     individual=True) #doctest: +SKIP

Reviewer 2: ```suggestion
    ...     individual=True)  # doctest: +SKIP
```

Reviewer 2: maybe we don't want to skip it but we can match the output with an ellipsis.

Author: Am I doing this correctly in https://github.com/scikit-learn/scikit-learn/pull/16619/commits/ef925fd148e68154b0cae7fa5f209f176e26fb71?

Author: Found it. Done in https://github.com/scikit-learn/scikit-learn/pull/16619/commits/c614425c060ae09ab05db2b47ef40e8dab5401fb

> You can access the newly created figure and Axes objects using ``plt.gcf()``

Reviewer 2: I would remove this part.

Indeed, `plot_partial_dependence` will return a `PartialDependenceDisplay` which contain the `ax_` attribute to make some matplotlib changes and you can use the same display to plot again the same information.

Author: Done in https://github.com/scikit-learn/scikit-learn/pull/16619/commits/5c8f0a2949254cc89f32484801991e8819cd4fbf

and ``plt.gca()``.

In ICE plots it might not be easy to see the average effect of the target
variable. Hence, it is recommended to use ICE plots alongside partial
dependency plots. They can be plotted together with ``individual='both'``.

    >>> plot_partial_dependence(clf, X, features,

>     ...     individual='both') #doctest: +SKIP

Reviewer 2: Maybe don't skip as well.

Author: Done in https://github.com/scikit-learn/scikit-learn/pull/16619/commits/c614425c060ae09ab05db2b47ef40e8dab5401fb

> Mathematical Definition

Reviewer 2: I feel this section a bit alone. Would it makes sense to have 

1. PDP
2. ICE
3. Mathematical defintion for PDP and ICE

@NicolasHug any thoughts?

Author: I have moved 'Mathematical definitions' and 'Computation method' sections to the bottom in https://github.com/scikit-learn/scikit-learn/pull/16619/commits/91258cae84afcd510b17cc1338d47411785b5806

Author: > Also we should update the paragraph on the method to explain that computing the individual conditional expectations is only possible when using the "brute" method. The faster "recursion" method is can only compute the average ICE values.
https://github.com/scikit-learn/scikit-learn/pull/16619/commits/a8a77f4400a7de1461498bf8ce94a620813e6715

^^^^^^^^^^^^^^^^^^^^^^^

> In ICE, for each instance in :math:`\{(X_S^{(i)}, X_C^{(i)})\}_{i=1}^N` the

Reviewer 2: ```suggestion
In ICE, for each sample in :math:`\{(X_S^{(i)}, X_C^{(i)})\}_{i=1}^N` the
```

:math:`x_C` for the features in :math:`X_C`. Note that :math:`x_S` and
:math:`x_C` may be tuples.

Computing this integral for various values of :math:`x_S` produces a PDP plot
as above.

In ICE, for each sample in :math:`\{(X_S^{(i)}, X_C^{(i)})\}_{i=1}^N` the
line :math:`pd_{X_S}^{(i)}` is calculated against :math:`X_S^{(i)}`, while

> :math:`X_C^{(i)}` remains fixed.

Reviewer: The use of the term "sample" is a bit confusing here. I would remove this part, and define ICEs below as suggested. Maybe we can just add

"The ICE values are defined below"

over the dataset `X` which is computationally intensive.

> 

Reviewer: I think the simplest way to introduce ICE is to add here something like:

```
Each of the f(x_S, x_C^{(i)}) corresponds to one ICE line evaluated at x_S.
Computing this for multiple values of x_S, one obtains a full ICE line.
As one can see, the average of the ICE lines correspond to the partial dependence line. 
```
